infrastructure/code/code.go

```diff
@@ -64,6 +86,41 @@ func (sc *Scanner) SupportedCommands() []snyk.CommandName {
 }
 
 func (sc *Scanner) Scan(ctx context.Context, _ string, folderPath string) []snyk.Issue {
+	// When starting a scan for a folderPath that's already scanned, the new scan will wait for the previous scan
+	// to finish before starting.
+	// When there's already a scan waiting, the function returns immediately with empty results.
+	waitForPreviousScan := false
+	scanStatus := NewScanStatus()
+	scanStatus.isRunning = true
+	sc.scanStatusMutex.Lock()
+	previousScanStatus, wasFound := sc.runningScans[folderPath]
+	if wasFound && previousScanStatus.isRunning {
+		if previousScanStatus.isPending {
+			return []snyk.Issue{}
+		}
+
+		waitForPreviousScan = true
+		scanStatus.isPending = true
+	}
+
+	sc.runningScans[folderPath] = scanStatus
+	sc.scanStatusMutex.Unlock()
+	if waitForPreviousScan {
+		<-previousScanStatus.finished // Block here until previous scan is finished
+
+		// Setting isPending = false allows for future scans to wait for the current
+		// scan to finish, instead of returning immediately
+		sc.scanStatusMutex.Lock()
+		scanStatus.isPending = false
+		sc.scanStatusMutex.Unlock()
+	}
+	defer func() {
+		sc.scanStatusMutex.Lock()
+		scanStatus.isRunning = false
+		close(scanStatus.finished)
+		sc.scanStatusMutex.Unlock()
+	}()
+
 	startTime := time.Now()
 	span := sc.BundleUploader.instrumentor.StartSpan(ctx, "code.ScanWorkspace")
```

Comment:
shouldn't this be the first thing done? Else the milliseconds spend before are ignored, which is not the semantics of the performance measurement.

Reply:
The problem is that on line 109 we block execution until the previous scan finished, so that's probably not a time we want to measure